submariner/pluggable-network-plugin-and-ovn.md

```diff
@@ -0,0 +1,342 @@
+## Add support for OVN with pluggable network-plugin support
+
+[submariner#778](https://github.com/submariner-io/submariner/issues/778)
+
+## Summary
+
+Today Submariner supports working with existing network-plugins which are based
+on the kube-proxy with iptables backend. Any network-plugin outside of this
+case won't work, OVN-Kubernetes is an example.
+
+This is a proposal to support pluggable network drivers, starting with OVN-Kubernetes.
+
+## Proposal
+
+Create `submariner-network-plugin-sync` which will be conditionally deployed by the
+`submariner-operator` for network plugins which require it. In the case of OVN it will be
+the microservice responsible for synchronizing state with the OVN NBDB. Another example
+of network-plugin which could make use of this service in the future would be Calico for
+automatic [handling of the IPPool entries](https://submariner.io/deployment/calico/)
+
+Re-architect Submariner route-agent in a way such that different network-plugin
+implementations can be supported. The re-architecture should create a re-usable library
+which other microservices including the new `submariner-network-plugin-sync` could
+make use of for handling common submariner events.
+
+Those drivers would be notified of different events like:
+
+* `Initialization` event
+
+    Called once during startup, to let the plugin configure any necessary settings on the
+    system once the route-agent is ready
+
+* `Stop` event
+
+    Called once during stop to let the network plugin cleanup anything necessary during
+    at stop if at all.
+
+* `Uninstall` event
+
+    Called once if the route agent is being removed from a node, being Submariner un-installed,
+    or the route-agent driver being switched, a full cleanup will be necessary in this case.
+
+* `Transition to Non-Gateway` event
+
+    During Gateway to Non-Gateway events on the specific node may need to perform specific
+    cleanups
+
+* `Transition to Gateway` event
+
+    When a node becomes Gateway some settings may need to be adjusted on the node
+    by the driver.
+
+* `Periodic` event
+
+   We could provide a periodic event to the driver, but may be an alternative
+   is doing this directly in the driver, using a timer.
+
+* `Creation of a local Endpoint`
+
+* `Creation of a remote Endpoint`
+
+* `Update of a local Endpoint`
+
+* `Update of a remote Endpoint`
+
+* `Deletion of a local Endpoint`
+
+* `Deletion of a remote Endpoint`
+
+* `Node creation` on the cluster
+
+* `Node removal` on the cluster
```

Comment:
Are these relevant for us?

Reply:
Yes, for both the route agent and the sync agent. Now the route agent listen on pods, but just to find Node ips. We plan to change that.